mantidproject/mantid
# Indirect -  Save Custom Detector Grouping to XML

**Report to:** @SpencerHowells

**Description of work.**
This PR adds:
- The ability to create a grouping workspace using a custom grouping string. To do this `CreateGroupingWorkspace` now has the `CustomGroupingString` property. This can create a grouping workspace based on a string such as `3+4+5,6-10,11:15,16`. This is the same syntax used for the `GroupDetectors` algorithm.
- The ability to save a custom grouping string from the Indirect Data reduction interface to an `.xml` file.
- Added the ability to load `.xml` grouping files into indirect data reduction.

It was decided not to add the ability to save detector groupings to a `.map` file. It is better to use `.xml` because:
- It is the standard within Mantid, and is therefore better tested and used by many scientific areas.
- It can do everything a `.map` file can do (and is arguably a simpler format when you get used to it).
- `.map` files are *not* being used in external programs by Indirect, and so `.xml` are just as useful.
- Better documentation on `.xml` format. Used widely in many different algorithms.

**To test:**
1. Indirect Data Reduction interface
2. Select `IRIS` and `graphite` as analyser.
2. Select `Custom` grouping in the grouping combobox
3. Enter a grouping string in the line edit. Try `3+4,6-25,30-50,51:53`
4. Click `Save Grouping`
5. Choose a file name and make sure it ends with `.xml`. Save it.
6. Open the newly created file.
7. Should look like this file:
[iris_detector_grouping_test.zip](https://github.com/mantidproject/mantid/files/5688548/iris_detector_grouping_test.zip)

8. Load this file into Data Reduction from the `File` grouping option.
9. Perform a reduction for `26176` run
10. Show data for the created workspace. It should have 6 spectra.

Closes #29984

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
